GPL/EventProbe/PathResolver.h

```diff
@@ -0,0 +1,180 @@
+ * so we make a best effort with a bounded loop, truncating particularly long
+ * paths.
+ */
+
+#ifndef EBPF_EVENTPROBE_PATHRESOLVER_H
+#define EBPF_EVENTPROBE_PATHRESOLVER_H
+
+#include "vmlinux.h"
+
+#include "Helpers.h"
+
+#define PATH_MAX 4096
+#define PATH_MAX_BOUND 4095
+
+// Maximum number of path components we'll grab before we give up and just
+// prepend "./" to the path. Even though bumping this won't increase the number
+// of instructions in the resolver (loops are not unrolled), it will increase
+// the number of states that need to be explored by the verifier, which, if
+// high enough, will bump up against BPF_COMPLEXITY_LIMIT_INSNS in the kernel
+// (set to 1,000,000 as of 5.3).
+#define PATH_RESOLVER_MAX_COMPONENTS 100
+
+// Map used as a scratch area by the path resolver to store intermediate state
+// (dentry pointers). Indexed by CPU number.
+struct bpf_map_def SEC("maps") path_resolver_scratch_map = {
+    .type        = BPF_MAP_TYPE_ARRAY,
+    .max_entries = 128,
+    .key_size    = sizeof(uint32_t),
+    .value_size  = sizeof(struct dentry) * PATH_RESOLVER_MAX_COMPONENTS,
+};
+
+static void
+ebpf_resolve_path_to_string(char *buf, struct path *path, const struct task_struct *task)
+{
+    long size      = 0;
+    bool truncated = true;
+
+    // This function uses BPF_CORE_READ instead of raw kernel memory accesses
+    // throughout. This is due to the pointer arithmentic below with
+    // container_of (see note about struct mount below). BTF type information
+    // is lost when we use container_of to access the struct mount, so we can't
+    // use raw kernel memory dereferences.
+    //
+    // While we still have BTF information for task and other structs in this
+    // function, BPF_CORE_READ is used for all of them too for uniformity and
+    // simplicity.
+    struct fs_struct *fs_struct    = BPF_CORE_READ(task, fs);
+    struct path root               = BPF_CORE_READ(fs_struct, root);
+    struct vfsmount *curr_vfsmount = BPF_CORE_READ(path, mnt);
+
+    // All struct vfsmount's are stored in a struct mount. We need fields in
+    // the struct mount to continue the dentry walk when we hit the root of a
+    // mounted filesystem.
+    struct mount *mnt          = container_of(curr_vfsmount, struct mount, mnt);
+    struct dentry *curr_dentry = BPF_CORE_READ(path, dentry);
+    struct dentry **dentry_arr;
+
+    unsigned long cpu = bpf_get_smp_processor_id();
+    if (!(dentry_arr = bpf_map_lookup_elem(&path_resolver_scratch_map, &cpu))) {
+        ebpf_debug("Could not get path resolver scratch area for cpu %d", cpu);
+        return;
+    }
+
+    // Loop 1, follow the dentry chain (up to a maximum of
+    // PATH_RESOLVER_MAX_COMPONENTS) and store pointers to each dentry in
+    // dentry_arr
+    for (int i = 0; i < PATH_RESOLVER_MAX_COMPONENTS; i++) {
+        if (curr_dentry == root.dentry && curr_vfsmount == root.mnt) {
+            // We've reached the global root if either the current dentry or
```

Comment:
```suggestion
            // We've reached the global root if the current dentry and
```
right? (at least that is what the if logic is)

Reply:
Thanks for the catch. Yes, you're right. This mistake is a result of me puzzling the [boolean logic](https://elixir.bootlin.com/linux/latest/source/fs/d_path.c#L106) in Linux's own implementation out wrong when trying to describe it. Got my demorgans law wrong when trying to do it mentally.